File: sample-conference-videochat/sample-conference-videochat/Components/Users/ViewControllers/AddUsersViewController/AddUsersViewController.m
//
//  AddUsersViewController.m
//  sample-multiconference-videochat
//
//  Copyright (c) 2017 QuickBlox. All rights reserved.
//

#import "AddUsersViewController.h"
#import "UsersDataSource.h"
#import "SVProgressHUD.h"
#import "QBDataFetcher.h"

@interface AddUsersViewController ()

@property (strong, nonatomic) UsersDataSource *dataSource;

@end

@implementation AddUsersViewController

// MARK: Lifecycle

- (void)dealloc {
    
    ILog(@"%@ - %@",  NSStringFromSelector(_cmd), self);
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.dataSource = [UsersDataSource usersDataSource];
    NSMutableArray *users = [[NSMutableArray alloc] init];
    for (QBUUser *user in self.usersDataSource.objects) {
        if (![self.chatDialog.occupantIDs containsObject:@(user.ID)]) {
            [users addObject:user];
        }
    }
    self.dataSource.objects = [users copy];
    
    self.tableView.dataSource = self.dataSource;
    self.tableView.rowHeight = 44;
    
    // adding refresh control task
    if (self.refreshControl) {
        
        [self.refreshControl addTarget:self
                                action:@selector(fetchData)
                      forControlEvents:UIControlEventValueChanged];
    }
    
    UIBarButtonItem *createChatButton =
    [[UIBarButtonItem alloc] initWithTitle:@"Update"
                                     style:UIBarButtonItemStylePlain
                                    target:self
                                    action:@selector(didPressUpdateChatButton:)];
    
    self.navigationItem.rightBarButtonItem = createChatButton;
    self.navigationItem.rightBarButtonItem.enabled = NO;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if (self.refreshControl.refreshing) {
        [self.tableView setContentOffset:CGPointMake(0, -self.refreshControl.frame.size.height) animated:NO];
    }
}

- (void)didPressUpdateChatButton:(UIBarButtonItem *)item {
    
    [SVProgressHUD show];
    NSMutableArray *pushOccupantsIDs = [[NSMutableArray alloc] init];
    for (QBUUser *user in self.dataSource.selectedObjects) {
        [pushOccupantsIDs addObject:[NSString stringWithFormat:@"%tu", user.ID]];
    }
    self.chatDialog.pushOccupantsIDs = [pushOccupantsIDs copy];
    
    __weak __typeof(self)weakSelf = self;
    [QBRequest updateDialog:self.chatDialog successBlock:^(QBResponse * _Nonnull response, QBChatDialog * _Nullable chatDialog) {
        
        weakSelf.chatDialog.occupantIDs = chatDialog.occupantIDs;
        [SVProgressHUD dismiss];
        [weakSelf.navigationController popViewControllerAnimated:YES];
        
    } errorBlock:^(QBResponse * _Nonnull response) {
        
        [SVProgressHUD showErrorWithStatus:[NSString stringWithFormat:@"%@", response.error.reasons]];
    }];
}


// MARK: UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [self.dataSource selectObjectAtIndexPath:indexPath];
    [tableView reloadSections:[NSIndexSet indexSetWithIndex:0] withRowAnimation:UITableViewRowAnimationNone];
    
    self.navigationItem.rightBarButtonItem.enabled = self.dataSource.selectedObjects.count > 0;
}

// MARK: Private

- (void)fetchData {
    
    __weak __typeof(self)weakSelf = self;
    [QBDataFetcher fetchUsers:^(NSArray *users) {
        
        __typeof(weakSelf)strongSelf = weakSelf;
        NSMutableArray *mutableUsers = [users mutableCopy];
        for (QBUUser *user in users) {
            if ([strongSelf.chatDialog.occupantIDs containsObject:@(user.ID)]) {
                [mutableUsers removeObject:user];
            }
        }
        [strongSelf.dataSource setObjects:[mutableUsers copy]];
        [strongSelf.tableView reloadData];
        [strongSelf.refreshControl endRefreshing];
    }];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Users/ViewControllers/AddUsersViewController/AddUsersViewController.swift
//
//  AddUsersViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import SVProgressHUD

class AddUsersViewController: UITableViewController {
    // MARK: - Properties
    weak var usersDataSource: UsersDataSource?
    weak var chatDialog: QBChatDialog?
    private var dataSource = UsersDataSource()
    
    // MARK: Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        guard let usersDataSource = usersDataSource,
            let chatDialog = chatDialog,
            let occupantIDs = chatDialog.occupantIDs  else {
                return
        }
        let users = usersDataSource.objects.filter({
            occupantIDs.contains(NSNumber(value: $0.id)) == false
        })
        dataSource.updateObjects(users)
        tableView.dataSource = dataSource
        tableView.rowHeight = 44.0
        
        // adding refresh control task
        if let refreshControl = self.refreshControl {
            refreshControl.addTarget(self, action: #selector(fetchData), for: .valueChanged)
        }
        let createChatButton = UIBarButtonItem(title: "Update",
                                               style: .plain,
                                               target: self,
                                               action: #selector(didPressUpdateChatButton(_:)))
        
        navigationItem.rightBarButtonItem = createChatButton
        navigationItem.rightBarButtonItem?.isEnabled = false
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        if let refreshControl = self.refreshControl, refreshControl.isRefreshing == true {
            let contentOffset = CGPoint(x: 0, y: -refreshControl.frame.size.height)
            tableView.setContentOffset(contentOffset, animated: false)
        }
    }
    deinit {
        debugPrint("deinit \(self)")
    }
    
    //MARK: - Actions
    @objc func didPressUpdateChatButton(_ item: UIBarButtonItem?) {
        SVProgressHUD.show()
        var pushOccupantsIDs: [String] = []
        for user in dataSource.selectedObjects {
            pushOccupantsIDs.append(String(format: "%tu", user.id))
        }
        chatDialog?.pushOccupantsIDs = pushOccupantsIDs
        QBRequest.update(chatDialog!, successBlock: { [weak self] response, chatDialog in
            
            self?.chatDialog?.occupantIDs = chatDialog.occupantIDs
            SVProgressHUD.dismiss()
            self?.navigationController?.popViewController(animated: true)
            
            }, errorBlock: { response in
                SVProgressHUD.showError(withStatus: "\(String(describing: response.error?.reasons))")
        })
    }
    
    // MARK: UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        dataSource.selectObject(at: indexPath)
        tableView.reloadSections(NSIndexSet(index: 0) as IndexSet, with: .none)
        navigationItem.rightBarButtonItem?.isEnabled = dataSource.selectedObjects.count > 0
    }
    
    // MARK: - Internal Methods
    @objc private func fetchData() {
        QBDataFetcher.fetchUsers({ [weak self] users in
            guard let users = users else { return }
            let filteredUsers = users.filter({
                self?.chatDialog?.occupantIDs?.contains(NSNumber(value: $0.id)) == false
            })
            self?.dataSource.updateObjects(filteredUsers)
            self?.tableView.reloadData()
            self?.refreshControl?.endRefreshing()
        })
    }
}
